File: src/main/kotlin/philo/peanutbox/core/feature/classifier/ClassTypeClassifier.java
package philo.peanutbox.core.feature.classifier;

public abstract class ClassTypeClassifier implements TypeClassifier {

    public static boolean isSimpleConcreteClass(Class<?> clazz) {

        return isConcreteClass(clazz) && !hasInterfaces(clazz);
    }

    public static boolean isInterface(Class<?> clazz) {

        return clazz.isInterface();
    }

    public static boolean isConcreteClassImplemented(Class<?> clazz) {

        return isConcreteClass(clazz) && hasInterfaces(clazz);
    }

    private static boolean isConcreteClass(Class<?> clazz) {

        return !clazz.isInterface();
    }

    private static boolean hasInterfaces(Class<?> type) {

        return type.getInterfaces().length > 0;
    }
}

File: src/main/kotlin/philo/peanutbox/core/feature/classifier/ClassTypeClassifier.kt
package philo.peanutbox.core.feature.classifier

object ClassTypeClassifier : TypeClassifier {

    fun isSimpleConcreteClass(clazz: Class<*>): Boolean {
        return isConcreteClass(clazz) && hasInterfaces(clazz).not()
    }

    fun isInterface(clazz: Class<*>): Boolean {
        return clazz.isInterface
    }

    fun isConcreteClassImplemented(clazz: Class<*>): Boolean {
        return isConcreteClass(clazz) && hasInterfaces(clazz)
    }

    private fun isConcreteClass(clazz: Class<*>): Boolean {
        return isInterface(clazz).not()
    }

    private fun hasInterfaces(clazz: Class<*>): Boolean {
        return clazz.interfaces.isNotEmpty()
    }
}
